### PR TITLE
Added fees for distributions

### PR DESCRIPTION
**Pull request by Mingela**

### CODE REVIEW

File: brvs-core/src/main/java/iroha/validation/transactions/plugin/impl/SoraDistributionPluggableLogic.java [hunk 1]
```diff
@@ -327,11 +417,16 @@ private SoraDistributionProportions getSuppliesLeftAfterDistributions(
                   if (subtrahend == null) {
                     subtrahend = BigDecimal.ZERO;
                   }
-                  return entry.getValue().subtract(subtrahend);
+                  final BigDecimal feeSubtrahend = multiplyWithRespect(
+                      initialProportions.accountProportions.get(entry.getKey()),
+                      fee
+                  ).add(subtrahend);
+                  final BigDecimal remainder = entry.getValue().subtract(feeSubtrahend);
+                  return remainder.signum() == -1 ? BigDecimal.ZERO : remainder;
```
Comment (rkharisov) [inline brvs-core/src/main/java/iroha/validation/transactions/plugin/impl/SoraDistributionPluggableLogic.java]:
```suggestion
                  return entry.getValue().subtract(feeSubtrahend).max(BigDecimal.ZERO)
```

File: brvs-core/src/main/java/iroha/validation/transactions/plugin/impl/SoraDistributionPluggableLogic.java [hunk 2]
```diff
@@ -128,100 +181,119 @@ protected void applyInternal(Map<String, BigDecimal> processableObject) {
...
+      distributionTransactions.forEach((projectAccount, transactions) -> {
+        if (transactions != null && !transactions.isEmpty()) {
+          final Iterable<Transaction> atomicBatch = Utils.createTxAtomicBatch(
+              transactions,
+              brvsKeypair
+          );
+          final IrohaAPI irohaAPI = queryAPI.getApi();
+          irohaAPI.transactionListSync(atomicBatch);
+          final byte[] byteHash = Utils.hash(atomicBatch.iterator().next());
+          final TxStatus txStatus = trackHashWithLastResponseWaiting(irohaAPI, byteHash)
```
Comment (rkharisov) [inline brvs-core/src/main/java/iroha/validation/transactions/plugin/impl/SoraDistributionPluggableLogic.java:290]: What happens if the service is shutdown before the final transaction status is received?

File: brvs-core/src/main/java/iroha/validation/transactions/plugin/impl/SoraDistributionPluggableLogic.java [hunk 3]
```diff
@@ -128,100 +181,119 @@ protected void applyInternal(Map<String, BigDecimal> processableObject) {
...
-                  Entry::getKey,
-                  entry -> calculateAmountForDistribution(
-                      entry.getValue(),
-                      transferAmount,
-                      finalSuppliesLeft.accountProportions.get(entry.getKey())
-                  )
-              )
+        if (distributionFinished != null
+            && distributionFinished.finished != null
+            && distributionFinished.finished) {
```
Comment (rkharisov) [inline brvs-core/src/main/java/iroha/validation/transactions/plugin/impl/SoraDistributionPluggableLogic.java]:
```suggestion
        val isFinished = Optional.ofNullable(distributionFinished)
            .map(SoraDistributionFinished::getFinished)
            .orElse(false);
        if (isFinished) {
```

File: brvs-core/src/main/java/iroha/validation/transactions/plugin/impl/SoraDistributionPluggableLogic.java [hunk 4]
```diff
@@ -128,100 +181,119 @@ protected void applyInternal(Map<String, BigDecimal> processableObject) {
...
+        SoraDistributionProportions suppliesLeft = queryProportionsForAccount(
+            projectOwnerAccountId,
+            brvsAccountId
+        );
+        final SoraDistributionProportions initialProportions = queryProportionsForAccount(
+            projectOwnerAccountId
+        );
+        if (initialProportions == null
+            || initialProportions.accountProportions == null
+            || initialProportions.accountProportions.isEmpty()) {
```
Comment (rkharisov) [inline brvs-core/src/main/java/iroha/validation/transactions/plugin/impl/SoraDistributionPluggableLogic.java]:
```suggestion
        val isProportionsEmpty= Optional.ofNullable(initialProportions)
            .map(p -> p.accountProportions)
            .map(Map::isEmpty)
            .orElse(true);
        if (isProportionsEmpty) {
```

File: brvs-core/src/main/java/iroha/validation/transactions/plugin/impl/SoraDistributionPluggableLogic.java [hunk 5]
```diff
@@ -128,100 +181,119 @@ protected void applyInternal(Map<String, BigDecimal> processableObject) {
...
-              toDistributeMap
-          )
-      );
-    });
-    sendDistributions(transactionList);
+          return;
+        }
+        // if brvs hasn't set values yet
+        if (suppliesLeft == null || suppliesLeft.accountProportions == null
+            || suppliesLeft.accountProportions.isEmpty()) {
```
Comment (rkharisov) [inline brvs-core/src/main/java/iroha/validation/transactions/plugin/impl/SoraDistributionPluggableLogic.java]: I suggest to  use Optional here too

File: brvs-core/src/main/java/iroha/validation/transactions/plugin/impl/SoraDistributionPluggableLogic.java (hunk 2 above)
Comment (Mingela) [reply brvs-core/src/main/java/iroha/validation/transactions/plugin/impl/SoraDistributionPluggableLogic.java:290]: I added creation time sync so far. Later we will persist the status of a distribution

**Review by rkharisov — APPROVED** [review]

**Pull request merged**